soc: esp32s3: add TWAI driver configuration

EricNRS opened the pull request:
Add TWAI configuration for CAN.

EricNRS committed:
soc: esp32s3: add TWAI driver configuration

Add TWAI configuration for CAN.

Signed-off-by: Eric Holmberg <eric.holmberg@northriversystems.co.nz>

EricNRS commented:
Tested by running both `tests/drivers/can/timing` and `tests/drivers/can/api` tests and also on real hardware for both send and receive.  I'm assuming those get automatically run as part of the automated tests, but not sure since this my first time through the PR process.

kartben commented:
Thanks a lot for the PR!

> [...].  I'm assuming those get automatically run as part of the automated tests, but not sure since this my first time through the PR process.

You would need to add "can" to https://github.com/zephyrproject-rtos/zephyr/blob/66018d5f0a3ce4b817726701ae6cb14a483cf5e0/boards/xtensa/esp32s3_devkitm/esp32s3_devkitm.yaml#L7 in a subsequent commit (and also update the documentation page for the devkit to add CAN/TWAI as a supported feature).

EricNRS committed:
boards: esp32s3_devkitm: update CAN support

Add CAN to supported list and update documentation.

Signed-off-by: Eric Holmberg <eric.holmberg@northriversystems.co.nz>

EricNRS commented:
> Edit: XIAO ESP32S3 just got added too
Less than an hour ago!  Will rebase and do the updates.

EricNRS committed:
boards: xiao_esp32s3: update CAN support

Add CAN to supported list and update documentation.

Signed-off-by: Eric Holmberg <eric.holmberg@northriversystems.co.nz>

henrikbrixandersen commented:
@EricNRS Thank you for the PR. Please add a Signed-off-by line to each commit (see https://docs.zephyrproject.org/latest/contribute/guidelines.html).

EricNRS commented:
> @LucasTambor, can you confirm whether we need to add an entry at `tests/drivers/can/api/testcase.yaml` also? something like:

I have updated the commits to add this for the `esp32s3_devkitm` and also add the signed-off mentioned by @henrikbrixandersen (and now have checkpatch in my pre-push hook).

However, I am seeing build issues here where the `tests/drivers/can/api/twai-enable.overlay` is not being applied.  I have posted a discussion to Discord here for this issue:  

https://discord.com/channels/720317445772017664/991025779993370654/1125585359133225030


I have not added additional configuration to `xiao_esp32s3` yet -- will get the esp32s3_devkitm board working first.

henrikbrixandersen commented on boards/xtensa/esp32s3_devkitm/esp32s3_devkitm.dts:
```diff
@@ -93,6 +93,13 @@
 };
 
+&twai {
+	/* requires external CAN transceiver or jumper on RX and TX pins for loopback testing */
```
You should be able to use the internal loopback support added in e229898caf23809810a7aff7fbc217f143c0e878 in an overlay where needed. This should void the need for an external loopback jumper.

henrikbrixandersen replied:
(not a blocker, by the way)

EricNRS replied:
I had some trouble getting this working.  I am testing with two different ESP32-S3 boards -- one with a CAN transceiver and one without.  Here are some test results of trying different combinations.

Could you comment on case 2 to see if I used the input-on-output correctly?

### Case 1 - Standard Pinctrl With Separate TX/RX pins

- Pass - ESP32-S3 with CAN Transceiver
- Fail (blocks in `test_filters_added_while_stopped`)  - ESP32-S3 without CAN Transceiver
- Pass - ESP32-S3 without CAN Transceiver and jumper from GPIO13 to GPIO14

Here is the pinctrl DT:
`&pinctrl {
	twai_default: twai_default {
		group1 {
			pinmux = <TWAI_TX_GPIO13>;
		};
                group2 {
			pinmux = <TWAI_RX_GPIO14>;
		};`

### Case 2 - Pinctrl With Input on Output Enabled
Without a CAN transceiver, if I make use of [https://github.com/zephyrproject-rtos/zephyr/commit/e229898caf23809810a7aff7fbc217f143c0e878 e229898], I only get a 52.51% pass rate.  I'm not sure what the issue is here at the moment.  I'm happy to put a scope on anything that you want me to look at.

- Pass - ESP32-S3 with CAN Transceiver
- Fail (blocks in `test_filters_added_while_stopped`)  - ESP32-S3 without CAN Transceiver
- Pass - ESP32-S3 without CAN Transceiver and jumper from GPIO13 to GPIO14

Here is the pinctrl DT:
`&pinctrl {
	twai_default: twai_default {
		group1 {
			pinmux = <TWAI_TX_GPIO13>;
                        input-enable;
		};
                group2 {
			pinmux = <TWAI_RX_GPIO14>;
                        output-enable;
		};`

### Case 3 - Pinctrl With Input on Output Enabled and ESP32 GPIO MUX Signals Crossed
This works and replaces an actual wire jumper.  It is ESP32 specific, but so is the TWAI, so this should be okay.
- Pass - ESP32-S3 with CAN Transceiver
- Pass - ESP32-S3 without CAN Transceiver
- Pass - ESP32-S3 without CAN Transceiver and jumper from GPIO13 to GPIO14

`
&pinctrl {
	twai_default: twai_default {
		group1 {
                        /* Create a virtual jumper connecting TWAI_TX and TWAI_RX */
			pinmux = <ESP32_PINMUX(13, ESP_TWAI_RX, ESP_TWAI_TX)>;
                };
	};
};`

EricNRS replied:
Marking this as resolved now that the proper loopback configuration has been updated per Lucas' comments in comment https://github.com/zephyrproject-rtos/zephyr/pull/59918#discussion_r1255875533

henrikbrixandersen commented on boards/xtensa/xiao_esp32s3/xiao_esp32s3.yaml:
```diff
@@ -9,6 +9,7 @@ supported:
   - uart
   - i2c
   - spi
+  - can
```
This looks incomplete. Pinctrl and DTS changes were left out. If you prefer, I can handle the Xiao ESP32S3 changes in a follow-up PR as I have access to that board.

EricNRS replied:
Yes, I was holding off on adding the rest of the Xiao ESP32-S3 support until I got the esp32s3_devkitm fixed and will then add that.  Haven't figured out the best way to flag that part of the commit as a WIP.  Will hopefully have some time tonight to work on this again.

henrikbrixandersen replied:
Best way would be to mark this PR as draft - or just submit what you have now, without the Xiao board changes.

EricNRS replied:
Thank you, I have converted this to a draft until I have some time to figure out why the test case isn't running correctly for the esp32s3_devkitm.

LucasTambor replied:
@EricNRS The timeout happens because of the way twister connect to the boards. It attaches to serial port before flashing the device. On esp boards the flashing needs to come first, but this not handled correctly in twister right now.

EricNRS replied:
@LucasTambor Thanks for confirm that is a tool issue and not something that I am doing.  I didn't find an open issue for this using keywords "twister esp", but I did find this one that is somewhat related:  https://github.com/zephyrproject-rtos/zephyr/issues/57120

I have tested an ESP32-S3 board that should be equivalent to the Xiao board using this command and verified it passes:
`west build -p -b xiao_esp32s3 -T tests/drivers/can/api/drivers.can.api.twai && west flash && west espressif monitor`

EricNRS replied:
@henrikbrixandersen - please take a look at the Xiao ESP32-S3 changes and if you are happy with them, mark this discussion as resolved.

EricNRS committed:
tests: drivers: can: api: add virtual CAN TX <-> RX Jumper

Remove the need for the can_transceiver fixture by assigning the CAN TX
and CAN RX lines to the same GPIO pin to allow for testing without a
transceiver or physical jumper.

Signed-off-by: Eric Holmberg <eric.holmberg@northriversystems.co.nz>

LucasTambor commented on tests/drivers/can/api/twai-enable.overlay:
```diff
@@ -4,6 +4,23 @@
+	twai_default: twai_default {
+		group1 {
+                        /* Create a virtual jumper connecting TWAI_TX and TWAI_RX */
+			pinmux = <ESP32_PINMUX(13, ESP_TWAI_RX, ESP_TWAI_TX)>;
```
You can create an internal loopback using using the `output-enable` on input pins and `input-enable` on output pins, such as:
```suggestion
twai_default: twai_default {
		group1 {
			pinmux = <TWAI_RX_GPIO5>;
			output-enable; /* Enable internal loopback */
		};
		group2 {
			pinmux = <TWAI_TX_GPIO5>;
			input-enable; /* Enable internal loopback */
		};
```

EricNRS replied:
Ah, that was the one combination that I didn't try in my test that I listed in https://github.com/zephyrproject-rtos/zephyr/pull/59918#discussion_r1253918757!  I was thinking that the same GPIO pin would end up overwriting the GPIO mux, but this does work.  Will have to do another read of the code in pinctrl_esp32.c to make sure I understand it.

EricNRS replied:
I have updated the commit to use your code snippet.

EricNRS replied:
I'll mark this as resolved now that I've updated the commit.  Let me know if the etiquette is to have the initial commenter handle marking as resolved.

henrikbrixandersen reviewed (approved):
Thank you!
